File: lab7/Dictionary.java
//------------------------------------------------------------------------------
//// lab7
//// Program: Dictionary.java
//// Author:  James Michael Vrionis
//// ID    :  JVrionis
//// Date  :  12-1-16
//// 
////----------------------------------------------------------------------------


public class Dictionary implements DictionaryInterface {
   
   // Private Inner Node
   private class Node {
      Pair item;
      Node left;
      Node right;

      Node(Pair x) {
         item = x;
         left = null;
         right = null;
      }
   }
   // Private Inner Pair
   private class Pair {
      String key;
      String value;

      Pair(String k, String v) {
         key = k;
         value = v;
      }
   }
   // Private Fields
   private Node root;
   private int numItems;

   // Constructor
   public Dictionary() {
      root = null;
      numItems = 0;
   }
   // Private Helper Functions ------------------------------------------

   // findKey()
   // returns the Node containing key k in the subtree rooted at R,
   // or returns Null if no such Node exists 
   private Node findKey(Node R, String k) {
      if (R == null || R.item.key.equals(k)) return R;
      if (R.item.key.compareToIgnoreCase(k) > 0) 
         return findKey(R.left, k);      
      else return findKey(R.right, k);
   }

   // findParent()
   // returns the parent of N in the subtree rooted at R,
   // or returns NULL if N is equal to R. (pre: R != NULL) 
   Node findParent(Node N, Node R) {
      Node P = null;
      if (N != R) {
         P = R;
         while(P.left != N && P.right != N) {
            if (N.item.key.compareToIgnoreCase(P.item.key) < 0)
               P = P.left;
            else
               P = P.right;
         }
      }
      return P;
   }

   // findLeftmost()
   // returns the leftmost Node in the subtree rooted at R, or NULL if R is NULL 
   Node findLeftmost(Node R) {
      Node L = R;
      if (L != null)
         while(L.left != null) L = L.left; 
         //for ( ; L.left != null; L = L.left);
         return L;
      }

   // printInOrder()
   // printes the (key, value) pairs belonging to the subtree rooted at R in order
   // of increasing keys to file pointed to by out 
   void printInOrder(Node R) {
      if (R != null) {
         printInOrder(R.left); 
         System.out.println(R.item.key + " " + R.item.value); 
         printInOrder(R.right);
      }
   }

   // deleteAll()
   // deletes all Nodes in the subtree rooted at N 
   void deleteAll(Node N) {
      if (N != null) {
         deleteAll(N.left);
         deleteAll(N.right);
      }
   }

   // ADT Operations ------------------------------------------
   

   // isEmpty()
   // pre: none
   // post: returns true if D is empty, false otherwise
   public boolean isEmpty() {
      return (numItems == 0);
   }

   // size()
   // pre: none 
   // post: returns the number of (key, value) pairs in D
   public int size() {
      return numItems;
   }

   // lookup()
   // pre: none 
   // returns the value v sucvh that (k, v) is in D, 
   // or returns null if no such value v exists
   public String lookup(String key) {
      Node N;
      N = findKey(root, key);
      // if true N=Null, if false 
      return (N == null ? null : N.item.value);
   }

   // insert()
   // pre: lookup(key)==null
   // post: !isEmpty(), size() is increased by one 
   // inserts new (key,value) pair into this Dictionary
   public void insert(String key, String value) throws DuplicateKeyException{
      Node N, A, B, Find;
      Find = findKey(root,key);
      if(Find !=null ) { // findKey(root,key) 
         throw new DuplicateKeyException("Dictionary Error: insert() cannot insert duplicate keys");
      }
      N = new Node(new Pair(key, value));
      B = null;
      A = root;
      while( A != null ) {
         B = A;
         if(A.item.key.compareToIgnoreCase(key)>0) A = A.left;
         else A = A.right;
      }

      if( B == null ) root = N;
      else if(B.item.key.compareToIgnoreCase(key)>0) B.left = N;
      else B.right = N;
      numItems++;
   }

   // delete()
   // pre: lookup(key)!=null 
   // deletes pair with the given key
   public void delete(String key) throws KeyNotFoundException{
      Node N, P, S;
      if(findKey(root, key)==null){
         throw new KeyNotFoundException("Dictionary Error: delete() cannot delete non-existent key");
      }
      N = findKey(root, key);
      if( N.left == null && N.right == null ) {
         if( N == root ){
            root = null;
         }else{
            P = findParent(N, root);           
          if( P.right == N ) P.right = null;
          else P.left = null;
         }
      }else if( N.right == null ) {
         if( N == root ){
            root = N.left;
         }else{
            P = findParent(N, root);
            if( P.right == N ) P.right = N.left;
            else P.left = N.left;
         }
      }else if( N.left == null ){
         if( N == root ){
            root = N.right;
         }else{
            P = findParent(N, root);
            if( P.right == N ) P.right = N.right;
            else P.left = N.right;
         }
      }else {  
         S = findLeftmost(N.right);
         N.item.key = S.item.key;
         N.item.value = S.item.value;
         P = findParent(S, N);
         if( P.right == S ) P.right = S.right;
         else P.left = S.right;
      }
      numItems--;
   }

   // makeEmpty()
   // pre: none 
   // post: resets to the empty state
   public void makeEmpty() {
      deleteAll(root);
      root = null;
      numItems = 0;
   }

   // toString()
   // overrides object's toString method 
   public String toString(){
      printInOrder(root);
      return "";
   }
}

File: pa5/Dictionary.c
//------------------------------------------------------------------------------------
// Pa5
// Program: Dictionary.c
// Author:  James Michael Vrionis
// ID    :  JVrionis
// Date  :  11-30-16
// 
//------------------------------------------------------------------------------------

#include <stdio.h>
#include <ctype.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>
#include "Dictionary.h"

const int tableSize = 101; // must be a prime number

// Node Object
typedef struct NodeObj {
   char* key;
   char* value;
   struct NodeObj* next;
} NodeObj;

// Node
typedef NodeObj* Node;

// newNode
// Node Type Constructor 
Node newNode(char* k, char* v) {
   Node N = malloc(sizeof(NodeObj)); // Node - NodeObj
   assert(N!=NULL);
   N->key = k;
   N->value = v;
   N->next = NULL;
   return(N);
}

// freeNode
// Free Node memory
void freeNode(Node* pN) {
   if (pN != NULL && *pN != NULL) {
      free(*pN);
      *pN = NULL;
   }
}

// DictionaryObj
// Create data type
typedef struct DictionaryObj {
   Node *table;
   int numItems;
} DictionaryObj;

// Dictionary
typedef struct DictionaryObj* Dictionary;

// Dictionary Type Constructor
Dictionary newDictionary() {
   Dictionary D = malloc(sizeof(DictionaryObj)); // Dictionary - DictionaryObj
   assert(D!=NULL);
   D->table = calloc(tableSize,sizeof(Node));
   assert(D->table !=NULL);
   D->numItems = 0;
   return D;
}

// Free Dictionarymemory
// Dictionary Type destructor
void freeDictionary(Dictionary* pD) {
   if (pD != NULL && *pD != NULL) {
      if (!isEmpty(*pD) ) makeEmpty(*pD);
      //makeEmpty(*pD);
      free((*pD)->table);
      free(*pD);
      *pD = NULL;
   }
}

// isEmpty()
// pre: none
// post: return 1 if S is empty, 0 otherwise
int isEmpty(Dictionary D) {
   if( D==NULL) {
      fprintf(stderr, "Dictionary Error: calling isEmpty() on NULL Dictionary reference\n");
      exit(EXIT_FAILURE);
   }
   if(D->numItems>0) {
      return 0;
   }
   return 1;
}

// rotate_left()
// bits rotated in an unsigned int
unsigned int rotate_left(unsigned int value, int shift) {
   int sizeInBits = 8*sizeof(unsigned int);
   shift = shift & (sizeInBits - 1); //bitwise & 
   if ( shift == 0 ) {
      return value;
   }
   else {
      return (value << shift) | (value >> (sizeInBits - shift));
   }
}  

// pre_hash()
// Turn string into an unsigned int
unsigned int pre_hash(char* input) {
   unsigned int result = 0xBAE86554;
   while (*input) {
      result ^= *input++;
      result = rotate_left(result, 5);
   }  
   return result;
}

// hash()
// changes a string into an int within the range 0-tableSize-1
int hash(char* key) {
   return pre_hash(key)%tableSize;
}

// makeEmpty()
// pre: none
// post: Resets D to the empty state
void makeEmpty(Dictionary D) {
   Node N;
   if(D==NULL)
   {
      fprintf(stderr,"Dictionary Error: calling makeEmpty() on NULL Dictionary Reference");
      exit(EXIT_FAILURE);
   }
   for(int i = 0; i<tableSize; i++) {
      while(D->table[i] != NULL){
         N = D-> table[i];
         D->table[i]=N->next;
         freeNode(&N);
         N = NULL;
      }
   }
   D->numItems = 0; 
}

// size()
// pre: none
// post: returns # of (key,value) pairs in D
int size(Dictionary D) {
   return D->numItems;
}

// findKey()
// post: returns reference to Node with key in this dictionary
Node findKey(Dictionary D, char* k) {
   Node N;
   if( D==NULL ) {
      fprintf(stderr, "Dictionary Error: calling findKey() on NULL Dictionary reference\n");
      exit(EXIT_FAILURE);
   }
   N = D->table[hash(k)];
   while(N != NULL) {
      if(strcmp(N->key,k)== 0)
         break;
      N = N->next;
   }
   return N;
}

// lookup()
// pre: none
// post: returns value v s.t. (k,v) is in D, otherwise NULL if no such value exists
char* lookup(Dictionary D, char* k) {
   Node N;
   if( D == NULL ) {
      fprintf(stderr, "Dictionary Error: calling lookup() on NULL Dictionary\n");
      exit(EXIT_FAILURE);
   }
   N = findKey(D,k); // using findKey helper function
   return (N == NULL ? NULL : N ->value);

   /*
   if(D->numItems == 0) {
      fprintf(stderr, "Dictionary Error: calling lookup() on empty Dictionary reference");
      exit(EXIT_FAILURE);
   }
   if(findKey(D,k) == NULL)
      return NULL;
   else
      return findKey(D,k)->value;*/
}

// insert()
// pre: lookup(D,k) != NULL
// post: inserts new(k,v) pair into D at head
void insert(Dictionary D, char* k, char* v) {
   Node N, F;
   F = findKey(D,k);
   //int h = hash(k);
   if(D == NULL) {
      fprintf(stderr, "Dictionary Error: calling insert on NULL Dictionary\n");
      exit(EXIT_FAILURE);
   }
   if(F != NULL) {
      fprintf(stderr, "Dictionary Error: calling insert() on a pre-existing key");
      exit(EXIT_FAILURE);
   } 
   
   int h = hash(k);
   if(D->table[h] == NULL) {
      D->table[h]  = newNode(k, v);
      D->numItems++; 
   }else {
      N = newNode(k, v);
      N->next = D->table[h];
      D->table[h] = N;
      D->numItems++;
   }
}

// delete()
// pre: lookup(D,k) != NULL
// post: deletes pair with the key k
void delete(Dictionary D, char* k) {
   Node N,A;
   N = findKey(D,k);
   
   if( D == NULL ) {
      fprintf(stderr, "Dictionary Error: Calling delete() on NULL Dictionary Reference\n");
      exit(EXIT_FAILURE);
   }
   if (D->numItems == 0 ) {
      fprintf(stderr, "Dictionary Error: cannot delete() empty Dictionary\n");
      exit(EXIT_FAILURE);
   }
   if(N == NULL) {
   //if(findKey(D,k) == NULL) {
      fprintf(stderr, "Dictionary Error: cannot delete() non-existent key \n");
      exit(EXIT_FAILURE);
   }
      
   // Deletes head with only one value in table slot
   if(N == D->table[hash(k)] && N->next == NULL) {
      freeNode(&N);
      D->table[hash(k)] = NULL;
   // Delete head > one value in table slot   
   }else if(N == D->table[hash(k)]) {      
      D->table[hash(k)] = N->next;
      N = NULL;
      freeNode(&N);
   }else {
      A = D->table[hash(k)];
      //walk list with A until we reach N and replace N with A.
      while(A->next != N) {
         A = A->next;
      }
      A->next = N->next;
      freeNode(&N);
   }
   D->numItems--;

}

// printDictionary()
// pre: none
// post: prints text representation of D to file out pointed to
void printDictionary(FILE* out, Dictionary D) {
   Node N;
   if( D==NULL ) {
      fprintf(stderr, "Dictionary Error: calling printDictionary() on NULL Dictionary reference\n");
      exit(EXIT_FAILURE);
   }
   // Search through 0 till tableSize
   for(int i = 0; i < tableSize; i++) {
      N = D->table[i];
   
      // while current node is not free
      while(N != NULL) {
         fprintf(out, "%s %s \n" , N->key, N->value);
         N = N->next;
      }
   } 
}
